### PR TITLE
Bump swig/4.1.0

### LINKED ISSUE #13741: [request] swig/4.1.0
### Package Details
  * Package Name/Version: **swig/4.1.0**
  * https://github.com/swig/swig/releases/tag/v4.1.0

The above mentioned version is newly released by the upstream project and not yet available as a recipe. Please add this version.

### PR DESCRIPTION
**Pull request by jmarrec**
Specify library name and version:  **swig/4.1.0**

* Fix #13741
* Add recipe for SWIG 4.1.0
* Note: The 0001 patch fails for 4.1.0, so make a specific one.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/swig/all/conanfile.py [hunk 1]
```diff
@@ -26,8 +26,16 @@ def _source_subfolder(self):
...
+    def _use_pcre2(self):
+        return self.version not in ['4.0.1', '4.0.2']
+
+
     def requirements(self):
-        self.requires("pcre/8.45")
+        if self._use_pcre2:
+            self.requires("pcre2/10.40")
+        else:
+            self.requires("pcre/8.45")
```
Comment (jmarrec) [inline recipes/swig/all/conanfile.py:38]: SWIG 4.1.0 uses pcre2, not pcre.

File: recipes/swig/all/conanfile.py [hunk 2]
```diff
@@ -83,8 +91,8 @@ def _configure_autotools(self):
 
         libargs = list("-L\"{}\"".format(p) for p in deps_libpaths) + list("-l\"{}\"".format(l) for l in deps_libs)
         args = [
-            "PCRE_LIBS={}".format(" ".join(libargs)),
-            "PCRE_CPPFLAGS={}".format(" ".join("-D{}".format(define) for define in deps_defines)),
+            "{}_LIBS={}".format("PCRE2" if self._use_pcre2 else "PCRE", " ".join(libargs)),
+            "{}_CPPFLAGS={}".format("PCRE2" if self._use_pcre2 else "PCRE", " ".join("-D{}".format(define) for define in deps_defines)),
```
Comment (jmarrec) [inline recipes/swig/all/conanfile.py:95]: Pass the right flags.

Note that I think swig has always used `PCRE_CFLAGS` (now `PCRE2_CFLAGS`), not `PCRE_CPPFLAGS`

cf https://github.com/swig/swig/blob/efe5f181cf69975cf2fa1dcddefc29b93e4a6196/configure.ac#L71

File: recipes/swig/all/patches/0002-4.1.0-do-not-define-SWIG_LIB_WIN_UNIX.patch [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
++++ configure.ac
+@@ -2825,7 +2825,7 @@ case $build in
+   *-*-cygwin*) SWIG_LIB_WIN_UNIX=`cygpath --mixed "$SWIG_LIB"`;;
+   *) SWIG_LIB_WIN_UNIX="";;
+ esac
+-AC_DEFINE_UNQUOTED(SWIG_LIB_WIN_UNIX, ["$SWIG_LIB_WIN_UNIX"], [Directory for SWIG system-independent libraries (Unix install on native Windows)])
++AC_DEFINE_UNQUOTED(SWIG_LIB_WIN_UNIX, [""], [Directory for SWIG system-independent libraries (Unix install on native Windows)])
+ 
+ SWIG_LIB_PREINST=$ABS_SRCDIR/Lib
+ AC_SUBST(SWIG_LIB_PREINST)
```
Comment (jmarrec) [inline recipes/swig/all/patches/0002-4.1.0-do-not-define-SWIG_LIB_WIN_UNIX.patch:11]: Another thing worth mentioning is that as of 4.1.0, SWIG has a CMakeLists.txt available for building.

File: recipes/swig/all/conandata.yml [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
+    sha256: "5b9313b1af5edfcea158a389520be266f013bc9be4ce933d79a30c5659ba99fe"
   "4.0.2":
     url: "https://github.com/swig/swig/archive/rel-4.0.2.tar.gz"
     sha256: "81d7ce78371f378a3299ddc5aea1da9a6178f325dcabb695d1b742f9e24a0fa6"
   "4.0.1":
     url: "https://github.com/swig/swig/archive/rel-4.0.1.tar.gz"
     sha256: "2eaf6fb89d071d1be280bf995c63360b3729860c0da64948123b5d7e4cfb6cb7"
 patches:
+  "4.1.0":
+    - base_path: "source_subfolder"
```
Comment (prince-chrismc) [inline recipes/swig/all/conandata.yml:13]: we have new https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

If you one and close the linter was merged jsut recently to check them 😉

File: recipes/swig/all/conanfile.py [hunk 5]
```diff
@@ -26,17 +26,25 @@ def _source_subfolder(self):
     def _settings_build(self):
         return getattr(self, "settings_build", self.settings)
 
+    @property
+    def _use_pcre2(self):
+        return self.version not in ['4.0.1', '4.0.2']
```
Comment (prince-chrismc) [inline recipes/swig/all/conanfile.py:31]: Maybe? not that another version _should_ get added but heck better safe then sorry

```suggestion
        return tools.Version(self.version) >= "4.1.0"
```

File: recipes/swig/all/patches/0002-4.1.0-do-not-define-SWIG_LIB_WIN_UNIX.patch (hunk 3 above)
Comment (prince-chrismc) [reply recipes/swig/all/patches/0002-4.1.0-do-not-define-SWIG_LIB_WIN_UNIX.patch:11]: YES! 🚀

**Review by prince-chrismc — COMMENTED** [review]
Just the patch fields for the new version pretty please

**Review by prince-chrismc — APPROVED** [review]
Sorry this one got lost in my inbox.

Totally fine by me

**Pull request merged**